### PR TITLE
Add bin/unbin support

### LINKED ISSUE #25422: Add functions `bin` and `unbin`
**Use case**

Viewing binary data such as bitmasks, hashes, internal representation of IEEE-754 numbers, etc.
Mostly for debugging.

**Describe the solution you'd like**

Similar to `hex`/`unhex`.

`bin` - returns a string with `0` and `1`. Bits are in "human" order, most significant first. Most significant zero bits are not printed (except the case with single zero).

**Describe alternatives you've considered**

Now I have to run interactive `python` from command line, but I will be more happy using ClickHouse.

**Caveats**

The name `bin` may suggest that the function will "split data into bins" but it is completely unrelated.

### PR DESCRIPTION
**Pull request by zxc111**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add bin/unbin functions support

close #25422

### CODE REVIEW

File: src/Functions/FunctionsCoding.h [hunk 1]
```diff
@@ -1326,6 +1326,395 @@ class FunctionUnhex : public IFunction
...
+        return std::make_shared<DataTypeString>();
+    }
+
+    static void unbinOne(const char * pos, const char * end, char *& out)
+    {
+        uint8_t left = 0;
+        for (int left_cnt = (end - pos) & 7; left_cnt > 0; --left_cnt)
+        {
+            left = left << 1;
+            if (*pos == '1')
```
Comment (vdimir) [inline src/Functions/FunctionsCoding.h:1639]: We consider all non `1` chars as zero. I thought that we can throw error if input not valid (contains anything else `0` or `1`), but `unhex` also process unknown symbols, so it's ok.

But we can make this "unspecified" behavior a bit clearer if all non zero chars will be treated as `1`.
Like `012` now is equivalent to `010`, but `011` seems more logic.

What do you think?

File: src/Functions/FunctionsCoding.h [hunk 2]
```diff
@@ -1326,6 +1326,395 @@ class FunctionUnhex : public IFunction
...
+            {
+                left += 1;
+            }
+            ++pos;
+        }
```
Comment (vdimir) [inline src/Functions/FunctionsCoding.h:1644]: Need a comment. What we are doing here, what `7` means, why we make such steps etc.

File: src/Functions/FunctionsCoding.h [hunk 3]
```diff
@@ -1326,6 +1326,395 @@ class FunctionUnhex : public IFunction
...
+        if (0 != left)
+        {
+            *out = left;
+            ++out;
+        }
+
+        while (end - pos != 0)
+        {
+            int c = 0;
```
Comment (vdimir) [inline src/Functions/FunctionsCoding.h:1653]: Should it be byte, not int?
```suggestion
            UInt8 c = 0;
```

**Review by vdimir — CHANGES_REQUESTED** [review]
Thank you for you work!

Core implementation is good, I left some comments, I think most of them you will easily resolve.

One of them is suggestion to generalize implementation between bin and hex functions, I can help with it.

**Comment by zxc111** [comment]
@vdimir Thanks for review and suggest, I'm glad to resolve these comments.

File: tests/queries/0_stateless/01926_bin_unbin.reference [hunk 4]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (vdimir) [inline tests/queries/0_stateless/01926_bin_unbin.reference:1]: In your previous version it works correctly, `bin(0)` should return  `'0'`. I'll fix it.

File: src/Functions/FunctionsCoding.h [hunk 5]
```diff
@@ -1049,40 +1055,51 @@ class FunctionHex : public IFunction
...
+            if (getName() == "bin")
+            {
+                out_vec.resize((in_vec.size() - size) * word_size + size);
+            } else if (getName() == "hex")
+            {
+                out_vec.resize(in_vec.size() * word_size - size);
+            } else
+            {
+                throw Exception("new function is not implemented for " + getName(), ErrorCodes::NOT_IMPLEMENTED);
+            }
```
Comment (vdimir) [inline src/Functions/FunctionsCoding.h:1084]: It's the same case, isn't it? I'll fix it in my branch

File: src/Functions/FunctionsCoding.h [hunk 6]
```diff
@@ -1277,55 +1359,144 @@ class FunctionUnhex : public IFunction
...
-    bool useDefaultImplementationForConstants() const override { return true; }
-
-    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t /*input_rows_count*/) const override
+struct BinImpl
+{
+public:
+    static constexpr auto name = "bin";
+    static constexpr size_t word_size = 8;
+    template <typename T>
+    static void executeOneUInt(T x, char *& out)
```
Comment (vdimir) [inline src/Functions/FunctionsCoding.h:1370]: I suppose `writeBinByte` can be used here. See commit my branch.

File: src/Functions/FunctionsCoding.h [hunk 7]
```diff
@@ -1107,160 +1124,225 @@ class FunctionHex : public IFunction
...
         else
         {
             return false;
         }
     }
+};
 
-    static bool tryExecuteFixedString(const IColumn * col, ColumnPtr & col_res)
+template <typename Impl>
+class UnConversion : public IFunction
```
Comment (vdimir) [inline src/Functions/FunctionsCoding.h:1191]: Too generic name, I've changed it to `DecodeFromBinaryRepr`, but I'm not sure that it also good if you have any other ideas, you can snuggest it.

**Review by vdimir — COMMENTED** [review]
Now without copy-paste looks much better!


I made commits with minor style changes and fixes for some corner cases (I left comments for it too, but it fixed in my branch).
Also feel free to comment my changes.

I can't push it onto your PR. (I suppose [this thing](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) not enabled).
Let me push to PR, or you can merge my commits from my branch by youself.


See: https://github.com/ClickHouse/ClickHouse/tree/vdimir/function-bin-unbin

**Comment by zxc111** [comment]
thanks a lot.
I've rebased on you branch.

**Comment by vdimir** [comment]
@zxc111 

Oh, I broke build with my comment. let me push into this PR and I'll fix it not to ask rebase branch every time.

Also I looked to `hex` function always adds leading zero (like `hex(1) = '01'`), so we can add leading zeros to `bin` too, to be consistent. I'll make it too.

**Comment by zxc111** [comment]
@vdimir 
The option is enable already.
Anything else I need to enable?
Maybe you push incorrect repository, https://github.com/zxc111/ClickHouse-1 is right and https://github.com/zxc111/ClickHouse is golang driver. 😭
image

**Review by vdimir — APPROVED** [review]

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-11553.

**Pull request merged**